% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/withDoc.R
\name{getx1.biseq}
\alias{getx1.biseq}
\title{compute mixture data given projection matrix (bisulfite sequencing)}
\usage{
getx1.biseq(NB0 = "top_bonferroni", b0, methmat, sample.id, celltypes,
  MAXITER = 10000, x0 = NULL)
}
\arguments{
\item{NB0}{number of genes to be retained, ordered by f-stat p-value. Other options: "all" uses all genes, "top_bonferroni" uses genes with adjusted p-values <0.05 after bonferroni correction, "top_fdr" uses genes with adjusted p-values <0.05 after FDR correction. Default is "top-bonferroni"}

\item{b0}{output of \code{\link{getb0.biseq}}}

\item{methmat}{a matrix of counts with rows corresponding to methylation
sites and columns. Columns are: chr start   end strand coverage1 numCs1
numTs1 coverage2 numCs2 numTs2 ....}

\item{sample.id}{vector of sample IDs}

\item{celltypes}{vector of cell types}

\item{MAXITER}{integer number of iterations allowed}

\item{x0}{initial cell type composition for fitting}
}
\value{
x1 cell mixture of sample

converged convergence
}
\description{
compute mixture data given projection matrix (bisulfite sequencing)
}
