% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/withDoc.R
\name{prep_scrnaseq}
\alias{prep_scrnaseq}
\title{do quality control of scRNAseq based on library size, feature counts, chrM, cell cycle phase, and count threshold}
\usage{
prep_scrnaseq(scrna_mat, genenametype = "hgnc_symbol",
  cellcycle = NULL, count.threshold = NULL)
}
\arguments{
\item{scrna_mat}{count matrix  for single cell RNAseq}

\item{genenametype}{nomenclature for genes in count matrix: "hgnc_symbol" or "ensembl_id"}

\item{cellcycle}{filter for specific cell cycle phase: "G1", "G2M", or "S". Default is NULL, for no cell cycle phase filtering.}

\item{count.threshold}{remove genes where average count is less than count.threshold. Default is NULL, for no count threshold filtering}
}
\value{
count matrix after quality control
}
\description{
do quality control of scRNAseq based on library size, feature counts, chrM, cell cycle phase, and count threshold
}
